### ISSUE TITLE
7.69, 7.70 busy-loops when doing SOCKS connection

### ISSUE THREAD

**Issue by zloi-user** [curl/curl#5532]
curl loads the cpu, with many threads this is especially noticeable

on version 7.68.0 the problem is not observed

**Comment by bagder** [comment]
Then please give us a lot more details about the problem and how to reproduce!

**Comment by zloi-user** [comment]
sh script:
`(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
(curl --proxy socks4://182.30.226.133:8291 'https://www.google.com/favicon.ico' -o  -)&
`

**Comment by bagder** [comment]
That's not very helpful.

**Comment by zloi-user** [comment]
I have a libcurl-based proxy checker, when testing part of the bad proxies 7.69 and 7.70 - it loads the system 100%, on version 7.68 only ~2%

sorry for my english, used google translator

**Comment by jay** [comment]
Can you reproduce without forking? Is there a specific proxy connection causing the problem?

**Comment by zloi-user** [comment]
> Can you reproduce without forking? Is there a specific proxy connection causing the problem?

try the proxy 115.88.138.251:4145, I have this proxy currently loads the cpu core 100%
```
curl --proxy 'socks4://115.88.138.251:4145' 'https://www.google.com/favicon.ico' -o - >>/dev/null
```

**Comment by jzakrzewski** [comment]
I've quickly tried it out. It doesn't hang each time but it happens
```
me@machine /tmp $ curl --proxy 'socks4://115.88.138.251:4145' 'https://www.google.com/favicon.ico'
Warning: Binary output can mess up your terminal. Use "--output -" to tell 
Warning: curl to output it to your terminal anyway, or consider "--output 
Warning: <FILE>" to save to a file.
me@machine /tmp $ curl --proxy 'socks4://115.88.138.251:4145' 'https://www.google.com/favicon.ico' -v
*   Trying 115.88.138.251:4145...
* SOCKS4 communication to www.google.com:443
* SOCKS4 connect to IPv4 216.58.215.228 (locally resolved)
^C
me@machine /tmp $ curl -V
curl 7.70.0 (x86_64-pc-linux-gnu) libcurl/7.70.0 OpenSSL/1.1.1g zlib/1.2.11 libssh2/1.9.0_DEV
Release-Date: 2020-04-29
Protocols: dict file ftp ftps http https imap imaps pop3 pop3s rtsp scp sftp smtp smtps tftp 
Features: HTTPS-proxy Largefile libz NTLM SSL TLS-SRP UnixSockets
me@machine /tmp $ 
```
During the second run I had 96% on one core used by curl.

**Comment by jay** [comment]
> if it does not load the processor, try to run several times

...

> I've quickly tried it out. It doesn't hang each time but it happens

I ran it a few times and I don't see anything unusual here. Sometimes the transfer succeeds and sometimes it doesn't due to RST (shown as "Failed receiving connect request ack").

curl 7.71.0-DEV (i386-pc-win32) libcurl/7.71.0-DEV OpenSSL/1.0.2t nghttp2/1.40.0

**Comment by bagder** [comment]
It is system agnostic and happens when the SOCKS proxy closes the connection while curl runs the SOCKS state machine. See my proposed fix in #5542.

**Comment by bagder** [comment]
@zloi-user: it would be awesome if you could try that patch and tell us if it fixes the problem you see!

**Comment by zloi-user** [comment]
compiled 7.69.1 with a [patch](https://github.com/curl/curl/commit/bf74460cea32a156e6cc58c557f1235513c03652.patch)

problem still exists :(

**Comment by bagder** [comment]
That was the second of the two commits, and the first one was the one that fixed the problem for me. You should rather merge both of them!

**Comment by bagder** [comment]
I can merge them into single  one, hang on.

**Comment by zloi-user** [comment]
compiled 7.69.1 with two patches
[1](https://github.com/curl/curl/commit/4d3bcad865e2f7edc5ce61a19d5c5113d08d55d0.patch),[2](https://github.com/curl/curl/commit/bf74460cea32a156e6cc58c557f1235513c03652.patch)

cheers, the problem does not occur  :)

**Comment by jay** [comment]
> It is system agnostic and happens when the SOCKS proxy closes the connection while curl runs the SOCKS state machine. See my proposed fix in #5542.

Both Windows and Linux recv return 0 repeatedly but then Windows after a second will return -1 WSAECONNABORTED which is why it doesn't busy loop.
